File: logistic_regression_v_2/logistic_regression.py
import pandas as pd
import random
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

def preprocess(data):

    # Identify categorical and numerical columns
    numerical_cols = data.select_dtypes(include=['int64', 'float64']).columns[:-1]

    # Scale numerical columns
    scaler = StandardScaler()
    data[numerical_cols] = scaler.fit_transform(data[numerical_cols])
    
    return data


def main():
    randint = random.randint(0, 1000)  # Set random seed for reproducibility
    # Load dataset
    data = pd.read_csv('../data/creditcard_2023.csv')
    target_col = 'Class'

    # Preprocess data
    data = preprocess(data)

    # Split features and target
    X = data.drop(columns=[target_col])  # Features
    y = data[target_col]  # Target variable
    
    # Train-test split
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=randint)

    # Load and train the model
    model = LogisticRegression(random_state=randint)
    model.fit(X_train, y_train)

    # Make predictions
    y_pred = model.predict(X_test)

    # Evaluate accuracy
    accuracy = accuracy_score(y_test, y_pred)
    print(f"Random Seed: {randint}")
    print(f"Accuracy: {accuracy}")

if __name__ == "__main__":
    main()

File: logistic_regression_v_2/logistic_regression.jl
using CSV
using DataFrames
using MLJ
using MLJLinearModels
using MLJBase: machine, fit!, transform, coerce
using DataFrames: DataFrame, dropmissing

function preprocess(data::DataFrame)
   
    y = data.Class
    data = data[:, Not(:Class)]
  
    # Normalize numerical columns
    stand = Standardizer(count=true)
    X = transform(fit!(machine(stand, data)), data)

    y = coerce(y, Multiclass)
    
    return X, y
end

function main()
    # Load data
    data = CSV.File("data/creditcard_2023.csv") |> DataFrame
    X, y = preprocess(data)

    # Split the data, 80% for training and 20% for testing, shuffles the data
    (X_train, X_test), (y_train, y_test) = partition((X, y), 0.8, shuffle=true, multi=true)

    LogisticClassifier = @load LogisticClassifier pkg=MLJLinearModels

    model = LogisticClassifier()

    mach = machine(model, X_train, y_train)
    fit!(mach)

    ŷ = predict(mach, X_test)
    ŷ_labels = mode.(ŷ) 

    accuracy = mean(ŷ_labels .== y_test)
    println("Accuracy: $accuracy")
end

main()